Add dimension handling interface for ingestion and segment creation

jon-wei opened the pull request:
Description
--------------
This PR is a trimmed down version of #2760 that introduces a generic dimension handling infrastructure used for Long/Float dimension handling. However, only String dimensions are currently supported, to keep the patch at a reviewable size. Actual support for Longs/Floats will come in a later PR.

At a high level, type-specific handling for dimensions in the IncrementalIndex, IndexMerger has been extracted into a new interface. There are 3 associated classes with the new interface:

DimensionHandler: Top level class, provides comparators and used to create the other 3 classes
DimensionIndexer: Handles logic related to in-memory ingestion of rows (IncrementalIndex)
DimensionMerger: Handles logic related to merging of segments (IndexMerger)

StringDimensionHandler contains the implementations of these classes for String dimensions.

Validation
--------------
As this PR significantly refactors the indexing and segment creation code, it has the following risks:
- Possible performance degradation
- Regression failures

To address these risks, the following validation strategies were used:

* Performance Degradation
Benchmarks measuring the performance of IncrementalIndex row adds, IncrementalIndex reads, and segment merge/persist indicate that this PR does not introduce performance degradations in affected areas.

* Regression failures
This PR passes the Druid unit tests and integration tests. The hadoop integration test (ITHadoopIndexTest) was run/verified manually.



-----

Benchmarks from a c4.8xlarge

patch
```
Benchmark                                      (rollup)  (rowsPerSegment)  (schema)  Mode  Cnt       Score       Error  Units
IncrementalIndexReadBenchmark.read                 true            750000     basic  avgt   50  334866.077 ±  5572.233  us/op
IncrementalIndexReadBenchmark.read                false            750000     basic  avgt   50  377260.561 ± 29860.209  us/op
IncrementalIndexReadBenchmark.readWithFilters      true            750000     basic  avgt   50  279812.716 ±  3370.998  us/op
IncrementalIndexReadBenchmark.readWithFilters     false            750000     basic  avgt   50  326096.952 ±  2870.442  us/op

Benchmark                        (rollup)  (rowsPerSegment)  (schema)  Mode  Cnt       Score      Error  Units
IndexIngestionBenchmark.addRows      true             75000     basic  avgt   50  384164.000 ± 1966.527  us/op
IndexIngestionBenchmark.addRows     false             75000     basic  avgt   50  399771.783 ± 2446.356  us/op

Benchmark                   ( numSegments)  (rollup)  (rowsPerSegment)  (schema)  Mode  Cnt        Score        Error  Units
IndexMergeBenchmark.merge                5      true             75000     basic  avgt   50  5448759.483 ± 159695.715  us/op
IndexMergeBenchmark.merge                5     false             75000     basic  avgt   50  5302718.895 ± 178994.287  us/op
IndexMergeBenchmark.mergeV9              5      true             75000     basic  avgt   50  4533349.734 ± 316551.870  us/op
IndexMergeBenchmark.mergeV9              5     false             75000     basic  avgt   50  4460183.171 ± 287253.950  us/op

Benchmark                        (rollup)  (rowsPerSegment)  (schema)  Mode  Cnt        Score        Error  Units
IndexPersistBenchmark.persist        true             75000     basic  avgt   50  1306846.208 ± 170797.750  us/op
IndexPersistBenchmark.persist       false             75000     basic  avgt   50  1347429.256 ± 189093.734  us/op
IndexPersistBenchmark.persistV9      true             75000     basic  avgt   50  1205900.253 ± 181631.734  us/op
IndexPersistBenchmark.persistV9     false             75000     basic  avgt   50  1210533.192 ± 149366.322  us/op
```

master
```
Benchmark                                      (rollup)  (rowsPerSegment)  (schema)  Mode  Cnt       Score       Error  Units
IncrementalIndexReadBenchmark.read                 true            750000     basic  avgt   50  338904.783 ±  9734.948  us/op
IncrementalIndexReadBenchmark.read                false            750000     basic  avgt   50  404326.322 ± 34762.840  us/op
IncrementalIndexReadBenchmark.readWithFilters      true            750000     basic  avgt   50  281539.062 ±  4068.100  us/op
IncrementalIndexReadBenchmark.readWithFilters     false            750000     basic  avgt   50  341720.741 ±  5290.862  us/op

Benchmark                        (rollup)  (rowsPerSegment)  (schema)  Mode  Cnt       Score      Error  Units
IndexIngestionBenchmark.addRows      true             75000     basic  avgt   50  386509.637 ± 2585.987  us/op
IndexIngestionBenchmark.addRows     false             75000     basic  avgt   50  399927.398 ± 2514.297  us/op

Benchmark                   ( numSegments)  (rollup)  (rowsPerSegment)  (schema)  Mode  Cnt        Score        Error  Units
IndexMergeBenchmark.merge                5      true             75000     basic  avgt   50  6430053.684 ± 298164.558  us/op
IndexMergeBenchmark.merge                5     false             75000     basic  avgt   50  6461107.881 ± 356905.476  us/op
IndexMergeBenchmark.mergeV9              5      true             75000     basic  avgt   50  5003605.414 ± 265774.052  us/op
IndexMergeBenchmark.mergeV9              5     false             75000     basic  avgt   50  5147752.762 ± 330098.556  us/op

Benchmark                        (rollup)  (rowsPerSegment)  (schema)  Mode  Cnt        Score        Error  Units
IndexPersistBenchmark.persist        true             75000     basic  avgt   50  1366495.093 ± 179354.995  us/op
IndexPersistBenchmark.persist       false             75000     basic  avgt   50  1414207.294 ± 208999.251  us/op
IndexPersistBenchmark.persistV9      true             75000     basic  avgt   50  1297002.648 ± 133262.718  us/op
IndexPersistBenchmark.persistV9     false             75000     basic  avgt   50  1343949.275 ± 167867.933  us/op
```

fjy commented on processing/src/main/java/io/druid/segment/DimensionMergerLegacy.java:
```diff
@@ -0,0 +1,64 @@
+ *
+ * NOTE: Remove this class when the legacy IndexMerger is deprecated and removed.
+ */
+public interface DimensionMergerLegacy extends DimensionMerger
```
should we add a deprecated annotation to this class?

fjy commented:
@jon-wei merge conflicts now

jon-wei commented:
rebasing, making adjustments based on new filtering changes

drcrallen commented:
Looks like a real error.

drcrallen commented:
@jon-wei added a discuss label. Please remove it when ready for next round of review/merge

jon-wei commented:
@drcrallen thanks, I've re-opened this now, benchmarks have been updated

acslk commented on processing/src/main/java/io/druid/segment/DimensionHandler.java:
```diff
@@ -0,0 +1,185 @@
+   *
+   * @return A new DimensionIndexer object.
+   */
+  public DimensionIndexer makeIndexer();
```
Should the DimensionIndexer here have the same <EncodedType, ActualType> as the dimension handler? If so, it would be nice to enforce it.

jon-wei replied:
yes, the types should be the same. I've added type parameters to makeIndexer()

gianm commented:
Sick, it looks like `IndexMergeBenchmark` accidentally got faster :)

jon-wei commented:
@gianm I haven't benchmarked this change in isolation, but I believe the IndexMerger speedup is from difference in implementation for single value rows here:

https://github.com/druid-io/druid/pull/3217/files#diff-4cb77a6cbc8f0534054954700125f6afL252
https://github.com/druid-io/druid/pull/3217/files#diff-aa6b53e2cf54c0d4132f9f1ead3f80c2R151

i.e., skipping the unneeded creation of an ArrayBasedIndexedInts and just making the single-value array directly

acslk commented on processing/src/main/java/io/druid/segment/IndexMerger.java:
```diff
@@ -607,14 +595,15 @@ public Metadata apply(IndexableAdapter input)
         if (mergedCapabilities == null) {
           mergedCapabilities = new ColumnCapabilitiesImpl();
-          mergedCapabilities.setType(ValueType.STRING);
+          mergedCapabilities.setType(null);
```
Is this line necessary? The default type is currently null.

jon-wei replied:
it was in an earlier incarnation of this patch, but not anymore, thanks

acslk commented on processing/src/main/java/io/druid/segment/IndexMerger.java:
```diff
@@ -684,94 +677,23 @@ public void close() throws IOException
-        dimHasNullFlags.add(dimHasNull);
-
-        FileOutputSupplier dimOut = new FileOutputSupplier(IndexIO.makeDimFile(v8OutDir, dimension), true);
+      ArrayList<FileOutputSupplier> dimOuts = Lists.newArrayListWithCapacity(mergedDimensions.size());
```
This should still be final

jon-wei replied:
restored "final"

acslk commented on processing/src/main/java/io/druid/segment/IndexMerger.java:
```diff
@@ -684,94 +677,23 @@ public void close() throws IOException
+            ioPeon,
+            dimCapabilities.get(i),
+            progress
+        );;
```
extra semicolon

jon-wei replied:
removed extra semicolon

jon-wei commented:
rebasing again...

jon-wei commented:
rebased and addressed comments from @acslk

jon-wei commented:
latest rebase changed the incrementalindex reading path slightly due to rebasing on top of #3344, re-ran IncrementalIndexReadBenchmark and performance hasn't changed:

```
Benchmark                                      (rollup)  (rowsPerSegment)  (schema)  Mode  Cnt       Score       Error  Units
IncrementalIndexReadBenchmark.read                 true            750000     basic  avgt   50  334866.077 ±  5572.233  us/op
IncrementalIndexReadBenchmark.read                false            750000     basic  avgt   50  377260.561 ± 29860.209  us/op
IncrementalIndexReadBenchmark.readWithFilters      true            750000     basic  avgt   50  279812.716 ±  3370.998  us/op
IncrementalIndexReadBenchmark.readWithFilters     false            750000     basic  avgt   50  326096.952 ±  2870.442  us/op
```

gianm commented on processing/src/main/java/io/druid/segment/DimensionHandler.java:
```diff
@@ -0,0 +1,185 @@
+public interface DimensionHandler<EncodedType extends Comparable<EncodedType>, ActualType extends Comparable<ActualType>>
+{
+  /**
+   * Get the name of the dimension associated with this handler.
```
This is the name of the column, I guess? (So for the indexing side it's the output name and for the querying side it's the input name?)

gianm replied:
The javadoc could use some clarification imo.

jon-wei replied:
Added note that this name is the output name of the column during indexing, and the input column name during querying.

gianm commented on processing/src/main/java/io/druid/segment/DimensionHandler.java:
```diff
@@ -0,0 +1,185 @@
+ * to be contained within a type-specific set of handler objects, simplifying processing classes
+ * such as IncrementalIndex and IndexMerger and allowing for abstracted development of additional dimension types.
+ *
+ * A dimension may have two representations, an encoded representation and a full representation.
```
s/full/actual/ ?

jon-wei replied:
Replaced 'full' with 'actual'

gianm commented on processing/src/main/java/io/druid/segment/DimensionIndexer.java:
```diff
@@ -0,0 +1,323 @@
+ * When encoding is present, there are two relevant orderings for the encoded values.
+ *
+ * 1.) Ordering based on encoded value's order of ingestion
+ * 2.) Ordering based on converted actual value
```
When in doubt between 1. and 1), use both :)

gianm commented on processing/src/main/java/io/druid/segment/DimensionIndexer.java:
```diff
@@ -0,0 +1,323 @@
+   *
+   * @return An array containing an encoded representation of the input row value.
+   */
+  public Object processSingleRowValToIndexKey(Object dimValues);
```
This returns Object because it might need to return `int[]` or `Object[]`, I guess?

gianm replied:
Could `processSingleRowValToIndexKey` and `processRowValsListToIndexKey` be collapsed into one function that handles both lists and non-lists? It looks like in both cases, an array is returned (`int[]` for the string dimension handler). And there's only one call site, which does this:

```
        Object dimsKey = (rawDimVal instanceof List) ?
                         indexer.processRowValsListToIndexKey(rawDimVal) :
                         indexer.processSingleRowValToIndexKey(rawDimVal);
```

That instanceof check could be done inside the DimensionIndexer. Is there a reason not to?

jon-wei replied:
I originally had this as one method, but split it after PR review in a previous version of this patch:

https://github.com/druid-io/druid/pull/2760#discussion_r61626272

gianm replied:
haha, rad. Hmm, I think one method still makes sense, since the call site is doing an instanceof check anyway…

gianm replied:
hmm, would it be workable to make the return type `EncodedArrayType` rather than `Object`?

So the specialization for String handler would be EncodedArray = int[], Encoded = Integer, Actual = String

gianm replied:
My main reason for this comment is trying to reduce the number of Objects in the interface.

jon-wei replied:
I've changed `Object` to `EncodedArrayType` across the DimensionIndexer/DimensionMerger, where applicable

gianm commented on processing/src/main/java/io/druid/segment/DimensionIndexer.java:
```diff
@@ -0,0 +1,323 @@
+   * @param unsortedIntermediateValue value to convert
+   * @return converted value
+   */
+  public EncodedType getSortedEncodedValueFromUnsorted(EncodedType unsortedIntermediateValue);
```
Part of the contract here should be that once you call getSortedEncodedValueFromUnsorted, getUnsortedEncodedValueFromSorted, or getSortedIndexedValues, you can't call any of the methods that process new values anymore.

gianm replied:
mm, nevermind, I guess it does work since `sortedLookup` is invalid. But, the sorted ids could change. So _that_ should be part of the contract (i.e. if you do getSorted -> processSingleRow -> getSorted again, then you might not get the same id back)

jon-wei replied:
Added info that writes will potentially invalidate previous reads that contain sorted encodings in the class header documentation

gianm commented on processing/src/main/java/io/druid/segment/DimensionIndexer.java:
```diff
@@ -0,0 +1,323 @@
+ * @param <ActualType> class of the actual values
+ *
+ */
+public interface DimensionIndexer<EncodedType extends Comparable<EncodedType>, ActualType extends Comparable<ActualType>>
```
Thread-safety expectations should be documented in the interface. Based on StringDimensionIndexer it looks like the expectation is this is not thread safe.

jon-wei replied:
Added a "Thread Safety" section to the class header documentation

gianm commented on processing/src/main/java/io/druid/segment/DimensionIndexer.java:
```diff
@@ -0,0 +1,323 @@
+   * @param rhs dimension value array from a TimeAndDims key
+   * @return comparison of the two arrays
+   */
+  public int compareTimeAndDimsKey(Object lhs, Object rhs);
```
These could be EncodedArrayType if we're doing that

gianm replied:
Also this isn't really comparing a TimeAndDimsKey, it's comparing an EncodedArray

gianm replied:
Similar comments with the other methods involving TimeAndDims in the name

jon-wei replied:
Changed `object` to `EncodedArrayType` and fixed method names

gianm commented on processing/src/main/java/io/druid/segment/DimensionIndexer.java:
```diff
@@ -0,0 +1,323 @@
+   * @param asList if true, return an array; if false, return a list
+   * @return list or array containing the actual values corresponding to the encoded values in the input array
+   */
+  public Object convertTimeAndDimsKeyToActualArrayOrList(Object key, boolean asList);
```
Hmm, it should be fine for this to be `Actual[] convertEncodedArrayToActualArray(EncodedArrayType key)`. If a caller wants a list, `Arrays.asList` is pretty cheap.

jon-wei replied:
That function actually can also return single values instead of array/list; I decided to keep it as is for now and update the documentation to indicate that it can return single values.

The reasoning being:
- The caller would have to do `instanceof int[]`, `instanceof float[]`, or use the methods from `java.lang.reflect.Array` which are slow, to determine whether to turn the array into a list if needed
- If this method always returned an array (no single values either), the caller would face the same type explosion when trying to retrieve values or have to use `java.lang.reflect.Array`

gianm commented on processing/src/main/java/io/druid/segment/DimensionIndexer.java:
```diff
@@ -0,0 +1,323 @@
+   * @param key dimension value array from a TimeAndDims key
+   * @return array containing the sorted encoded values corresponding to the unsorted encoded values in the input array
+   */
+  public Object convertTimeAndDimsKeyToSortedEncodedArray(Object key);
```
`EncodedArrayType convertUnsortedToSorted(EncodedArrayType key)`?

jon-wei replied:
renamed to `convertUnsortedEncodedArrayToSortedEncodedArray`

gianm commented on processing/src/main/java/io/druid/segment/DimensionHandler.java:
```diff
@@ -0,0 +1,185 @@
+ * @param <EncodedType> class of the encoded values
+ * @param <ActualType> class of the actual values
+ */
+public interface DimensionHandler<EncodedType extends Comparable<EncodedType>, ActualType extends Comparable<ActualType>>
```
Does this object have state? Is it thread safe? Can a single handler be used to handle many dimensions, or should each handler only be used once? (the javadocs should explain these)

jon-wei replied:
Add note that DimensionHandler implementations should be stateless and thus thread-safe

gianm commented on processing/src/main/java/io/druid/segment/DimensionHandler.java:
```diff
@@ -0,0 +1,185 @@
+   *
+   * @return A value transformation function.
+   */
+  public Function<Object, ActualType> getValueTypeTransformer();
```
This is not used, I guess it was rolled into indexer.process*

jon-wei replied:
removed this method from the interface

gianm commented on processing/src/main/java/io/druid/segment/DimensionHandler.java:
```diff
@@ -0,0 +1,185 @@
+   * @param dimVals Array of row values
+   * @return Size of dimVals
+   */
+  public int getLengthFromArrayObject(Object dimVals);
```
getEncodedArrayLength?

gianm replied:
btw, should this be on the Indexer? it seems not useful unless you have an Indexer.

jon-wei replied:
I kept this in the Handler because it would also be used during query time (not implemented in this PR), where a DimensionIndexer would not be available.

gianm commented on processing/src/main/java/io/druid/segment/DimensionHandler.java:
```diff
@@ -0,0 +1,185 @@
+   * @param rhs array of row values
+   * @return integer indicating comparison result of arrays
+   */
+  public int compareRowboatKey(Object lhs, Object rhs);
```
Could this be `int compareEncodedArrays(EncodedArrayType lhs, EncodedArrayType rhs)`?

Also are the encoded values expected to be sorted or unsorted values? Does it matter?

gianm commented on processing/src/main/java/io/druid/segment/DimensionHandler.java:
```diff
@@ -0,0 +1,185 @@
+
+
+  /**
+   * Given a Column and the index of the current row, retrieve a row as an array of values.
```
This sentence doesn't seem to be accurate

jon-wei replied:
Fixed documentation for `getSubColumn()`

gianm commented on processing/src/main/java/io/druid/segment/QueryableIndexIndexableAdapter.java:
```diff
@@ -230,7 +234,7 @@ public boolean hasNext()
               }
-              for (Object dimension : dictionaryEncodedColumns) {
+              for (Object dimension : columns) {
                 if (dimension instanceof Closeable) {
```
This is always true, right?

jon-wei replied:
got rid of unnecessary Closeable check

gianm commented on processing/src/main/java/io/druid/segment/StringDimensionHandler.java:
```diff
@@ -0,0 +1,1609 @@
+import java.util.List;
+import java.util.Map;
+
+public class StringDimensionHandler implements DimensionHandler<Integer, String>
```
I think this file'd be easier to understand if broken up into at least four different files: handler, indexer, mergerv9, mergerlegacy. None of those share state with each other and they are doing pretty different jobs.

jon-wei replied:
Split these into separate files

gianm commented on processing/src/main/java/io/druid/segment/StringDimensionHandler.java:
```diff
@@ -0,0 +1,1609 @@
+  }
+
+
+  public class StringDimensionMerger implements DimensionMergerV9
```
It looks like there's a lot of code copied/pasted between here and StringDimensionMergerLegacy. Is it possible to share that? Maybe one extends the other (like the IndexMergers)? Or maybe they both call into some shared static functions? Or maybe the Legacy one has a V9 one as a member variable, and delegates to it for various ops?

jon-wei replied:
I had kept them totally separate initially to allow for easier deletion of StringDimensionMergerLegacy when the legacy IndexMerger is deprecated and removed.

Restructured this so that:
- DimensionMerger is base class with common methods for both V9 and legacy
- DimensionMergerV9 has V9 specific methods
- DimensionMergerLegacy has legacy specific methods
- StringDImensionMergerV9 is the base class for String implementation, StringDimensionMergerLegacy extends StringDimensionMergerV9 and overrides as needed, and also implements DimensionMergerLegacy. StringDimensioMergerLegacy can still be easily removed from the codebase when needed.

gianm commented on processing/src/main/java/io/druid/segment/DimensionMerger.java:
```diff
@@ -0,0 +1,137 @@
+   * @param adapters List of adapters to be merged.
+   * @throws IOException
+   */
+  public void mergeValueMetadataAcrossSegments(List<IndexableAdapter> adapters) throws IOException;
```
`writeValueMetadata` is probably a decent name for this. When I first looked at the current name, I was thinking, "where do the merged values go?"

jon-wei replied:
renamed to `writeMergedValueMetadata`

gianm commented on processing/src/main/java/io/druid/segment/DimensionMerger.java:
```diff
@@ -0,0 +1,137 @@
+   * @param segmentRow A row from a segment to be converted to its representation within the merged sequence of rows.
+   * @param segmentIndexNumber Integer indicating which segment the row originated from.
+   */
+  public Object convertSegmentRowValuesToMergedRowValues(Object segmentRow, int segmentIndexNumber);
```
Is `segmentRow` an EncodedArrayType? If so let's type it as that.

jon-wei replied:
Changed this to accept and return EncodedArrayType instead of Object

gianm commented on processing/src/main/java/io/druid/segment/DimensionMerger.java:
```diff
@@ -0,0 +1,137 @@
+   * @param rowValues The row values to be added.
+   * @throws IOException
+   */
+  public void processMergedRow(Object rowValues) throws IOException;
```
EncodedArrayType?

jon-wei replied:
Changed this to accept EncodedArrayType

gianm commented on processing/src/main/java/io/druid/segment/DimensionMerger.java:
```diff
@@ -0,0 +1,137 @@
+   *
+   * @throws IOException
+   */
+  public void buildIndexes(List<IntBuffer> segmentRowNumConversions) throws IOException;
```
This writes indexes, right? If so let's put "write" in the name, like `writeIndexes`

gianm replied:
I think what I'm saying is that things that write files should have "write" in their names.

jon-wei replied:
renamed to writeIndexes

gianm commented on processing/src/main/java/io/druid/segment/DimensionMerger.java:
```diff
@@ -0,0 +1,137 @@
+   *
+   * @return true if this dimension can be excluded from the merged segment.
+   */
+  public boolean canSkip();
```
Is there any way to find out if we can hop or jump too?

jon-wei commented:
still addressing PR comments from @gianm, marking as WIP temporarily

jon-wei commented:
Addressed PR comments from @gianm

fjy commented on processing/src/main/java/io/druid/segment/DimensionHandler.java:
```diff
@@ -0,0 +1,181 @@
+   * @param rhs array of row values
+   * @return integer indicating comparison result of arrays
+   */
+  public int compareSortedEncodedArrays(EncodedTypeArray lhs, EncodedTypeArray rhs);
```
what is this used for? is it used for merging? If so, should it be in merger?

gianm replied:
maybe it's going to be needed at query time too? @jon-wei?

jon-wei replied:
It's used only in the Rowboat compareTo() function, it doesn't depend on any merger state so I put it on the DimensionHandler, I can move it if needed

gianm replied:
handler sounds fine to me as it's stateless.

drcrallen replied:
if `EncodedType` must be `Comparable` why does this need to be a special thing?

fjy commented on processing/src/main/java/io/druid/segment/DimensionHandler.java:
```diff
@@ -0,0 +1,181 @@
+   * @param currRow The index of the row to retrieve
+   * @return The row from "column" specified by "currRow", as an array of values
+   */
+  public Object getRowValueArrayFromColumn(Closeable column, int currRow);
```
i find this a little bit confusing, does this function only work with subColumns from the previous method? If so, shoudl we have a more specific type than just closeable?

gianm replied:
the javadoc says it must be "given a subcolumn from getSubColumn"

drcrallen replied:
unless its closed in the method, which is not indicated, then it should accept `Object`

drcrallen replied:
OR have stricter typing.

fjy commented on processing/src/main/java/io/druid/segment/DimensionIndexer.java:
```diff
@@ -0,0 +1,350 @@
+   *
+   * The value within the returned array should be encoded if applicable, i.e. as instances of EncodedType.
+   *
+   * NOTE: This function can change the internal state of the DimensionIndexer.
```
I think there should be a bit more highlighted and add a link to proper usage (as explained in the comments above)

fjy commented:
👍  on the interface changes even if my comments are not addressed

gianm commented:
👍  LGTM

gianm commented:
Anyone else interested in reviewing? We do have two +1s already, but just checking. This has been open for a while and would like to get it in so we can begin building on top of it. If not then I'll merge it in a bit.

drcrallen commented:
I'd like to do a once-over if there is still time.

gianm commented:
@drcrallen go for it; although please for @jon-wei's sake I'm hoping we can get this in before other things that might conflict with it (which is basically anything indexing related). He's already done countless rebases :)

drcrallen commented:
if I can't get it done before EOD Friday I forfeit my right to complain about it

drcrallen commented on processing/src/main/java/io/druid/segment/DimensionHandler.java:
```diff
@@ -0,0 +1,181 @@
+   * @param dimVals Array of row values
+   * @return Size of dimVals
+   */
+  public int getLengthFromEncodedArray(EncodedTypeArray dimVals);
```
This is so that `EncodedTypeArray` does not have to guarantee a `length()` or similar method?

jon-wei replied:
The caller would otherwise have to check and cast the array to the appropriate type, or it would have to use the slow performance reflection method (Array.getLength())

drcrallen commented on processing/src/main/java/io/druid/segment/DimensionHandler.java:
```diff
@@ -0,0 +1,181 @@
+ * @param <EncodedType> class of the encoded values
+ * @param <ActualType> class of the actual values
+ */
+public interface DimensionHandler<EncodedType extends Comparable<EncodedType>, EncodedTypeArray, ActualType extends Comparable<ActualType>>
```
Maybe its later in the docs, but I don't see anything here about why the types need to be `Comparable`

drcrallen replied:
Can you expand on that requirement a bit?

jon-wei replied:
I put the Comparable requirement there since a column used as a dimension needs to be sortable, I can elaborate on that in the docs.

drcrallen replied:
Mostly I'm curious because you can either have items which are by their nature Comparable, or have some other item in the chain take a Comparator as an argument. And it is not clear that one is preferable over the other.

jon-wei replied:
I think I'm in favor of having the values being Comparable inherently, since the basic types (strings and numbers) are Comparable, and for a potential complex type? an implementer would have to implement comparator logic anyway for the column type to be usable as a dimension, so might as well have it as part of the type itself.

Do you foresee a use case where requiring a Comparable type would be a hindrance?

jon-wei replied:
The comparison logic should all be contained within the type-specific Dimension* implementations though, so removing the Comparable requirement shouldn't be a problem either.

drcrallen commented on processing/src/main/java/io/druid/segment/DimensionHandlerUtil.java:
```diff
@@ -0,0 +1,45 @@
+{
+  private DimensionHandlerUtil() {}
+
+  public static DimensionHandler getHandlerFromCapabilities(String dimensionName, ColumnCapabilities capabilities)
```
This feels like a major pinch point for extensibility. Meaning there is no discovery capabilities here for non-default dimension handling.

would it be possible to either use a guice multi-map binding, or a Jackson factory method with something like 

`objectMapper.convertValue(ImmutableMap.of("type", capabilities.getType()), DimensionHandlerFactory.class).buildHandler(dimensionName, capabilities)`

?

jon-wei replied:
sure, I can look into that

jon-wei replied:
hm, it's not clear to me right now what the best approach is here for supporting dimension type extensions. 

My goal with this PR was just to allow the new interface to work with Strings. let's handle this in a follow-up PR that adds a Long implementation?

drcrallen commented on processing/src/main/java/io/druid/segment/DimensionIndexer.java:
```diff
@@ -0,0 +1,350 @@
+ *
+ * Thread Safety
+ * --------------------
+ * Each DimensionIndexer exists within the context of a single IncrementalIndex. Before IndexMerger.persist() is
```
I don't understand this restriction. A class that is "thread safe until something happens, then it is not thread safe" is not thread safe.

gianm replied:
I think the intent is not that it becomes non-thread-safe after persist is called, but rather:

- it is expected that a single persist thread is the only thing calling the methods that deal with getting sorted value; i.e. multithreaded persisting is not ok
- after that thread starts calling those methods, writes are no longer safe

jon-wei replied:
@drcrallen hm, maybe the wording of that section could be improved 

to add to @gianm's comment, I generally avoided changing existing behavior with this interface; this section is describing how in the existing usage:
- IncrementalIndex can have concurrent read/writes until it is persisted; these interactions only use the unsorted ID space
- After IndexMerger persist() starts, no more writes to the IncrementalIndex should occur, the merge/persist is handled by a single thread, and the sorted ID space is only relevant in this stage

drcrallen commented on processing/src/main/java/io/druid/segment/DimensionIndexer.java:
```diff
@@ -0,0 +1,350 @@
+ *
+ * Index merge and persist are currently handled by a single thread.
+ *
+ * After IndexMerger.persist() is called on an IncrementalIndex, any indexers associated with that IncrementalIndex
```
Does this mean that `IndexMerger.persist()` must have a proper memory barrier of some kind?

jon-wei replied:
IndexMerger.persist() is currently handled by a single thread, so I don't think this is needed for now

drcrallen replied:
If thread A adds an event, and thread B calls persist, There needs to be proper memory barriers in place on the data used in the persist so that the memory seen in thread B is consistent with that written by A.

I'm actually not certain if the current impl guarantees this properly.

jon-wei replied:
In the current implementation, the Runnable that runs IndexMerger.persist() is submitted to an Executor by the same thread that handled the adds. This is the case in IndexGeneratorJob and the Plumbers, so that should take care of memory consistency between the threads according to this page:

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/package-summary.html#MemoryVisibility

"Actions in a thread prior to the submission of a Runnable to an Executor happen-before its execution begins. Similarly for Callables submitted to an ExecutorService."

This patch keeps the current behavior at these thread boundaries

drcrallen commented on processing/src/main/java/io/druid/segment/IndexIO.java:
```diff
@@ -331,90 +334,147 @@ public static void validateRowValues(
               "Expected nulls, found %s and %s",
-              Arrays.toString(dim1Vals),
-              Arrays.toString(dim2Vals)
+              Arrays.toString((int[]) dim1Vals),
```
This breaks type relaxing

drcrallen commented on processing/src/main/java/io/druid/segment/IndexIO.java:
```diff
@@ -331,90 +334,147 @@ public static void validateRowValues(
+              );
+            }
+          } else if (dim2ValsLen == 0 && dim1ValsLen == 1) {
+            final String dimValName = dim1ValNames.get(Array.getInt(dim1Vals, 0));
```
This breaks type assumptions as well unless I'm missing something.

drcrallen commented on processing/src/main/java/io/druid/segment/IndexIO.java:
```diff
@@ -331,90 +334,147 @@ public static void validateRowValues(
     for (int i = 0; i < dims1.length; ++i) {
-      final int[] dim1Vals = dims1[i];
-      final int[] dim2Vals = dims2[i];
+      final Object dim1Vals = dims1[i];
```
Ok, I'm lost in the code now. Is this an `int[]` or not? here it seems to be relaxed on type, but everywhere else in this method it assumes casting to `int[]` is fine.

drcrallen replied:
ah, I think I see. there are checks for dictionary encoding all through here. That doesn't seem like a sustainable solution. Is it possible to have the checks be done in a way that allows the capabilities to control the type assumptions?

jon-wei replied:
I've moved the type-specific parts of this IndexIO validation section into a method on DimensionHandler

drcrallen commented on processing/src/main/java/io/druid/segment/IndexIO.java:
```diff
@@ -748,7 +808,7 @@ public int size()
             serdeficator.write(channel);
             channel.close();
-          } else if (filename.startsWith("met_")) {
+          } else if (filename.startsWith("met_") || filename.startsWith("numeric_dim_")) {
```
By this methodology, is every dimension type going to have a condition here?

Dimension name assumptions need revamped overall, see https://github.com/druid-io/druid/pull/3118 for related code.

I'm not convinced such a change would be in scope for this PR, but if you have any thoughts on ways to handle new dimension types in the future it would be nice to capture them for posterity.

jon-wei replied:
My thinking in this area was that the V8 segment format/legacy IndexMerger would be deprecated by the time we really have dimension types beyond long/float, i.e., this code would no longer be relevant when it's time to extend it further. (could take the stance that users who want extra dimension types must use buildV9directly)

So this was a quick way for IndexIO's convertV8toV9() to be able to distinguish between dimensions that use a dictionary encoded String style column, or ones that use the metric-style.

drcrallen replied:
Ok cool, could you slap a quick comment in the code that the approach is deprecated not intended to be used to add more dimension types, instead favoring the v9 approach or later?

jon-wei commented:
@drcrallen @gianm I moved the IndexIO type-specific validation into a method on DimensionHandler and updated parts of the javadocs in the new interfaces, can you take another look?

drcrallen commented on processing/src/main/java/io/druid/segment/StringDimensionMergerV9.java:
```diff
@@ -0,0 +1,575 @@
+
+    // write dim values to one single file because we need to read it
+    File dimValueFile = IndexIO.makeDimFile(outDir, dimensionName);
+    FileOutputStream fos = new FileOutputStream(dimValueFile);
```
these feel a little shaky without some sort of try block. Are these resources guaranteed to be cleaned up properly?

jon-wei replied:
this part was copied without revision from IndexMergerV9, I'm somewhat of wary of changing this behavior without more thought and testing, let's handle this in a follow-up PR

drcrallen commented:
@jon-wei thanks.

I'm out of time for review, so I ask one last thing. Please check the output of indexing something like the wiki set or TPCH data, and report back on if they are binary equivalent pre and post this PR, and if they are not, make sure it is well understood why they are not.

gianm commented:
👍  to the most recent changes. binary equivalence confirmed on wiki quickstart sample data:

```
18d4cf54a45961e30b264824e3757fbf912107fb  index/00000.smoosh
35406c92ad30b1d0cfa507fe9d9a48b14c4046ad  index/meta.smoosh
febea93b2c7d75eb566bf43de2488821733303f5  index/version.bin
18d4cf54a45961e30b264824e3757fbf912107fb  index2/00000.smoosh
35406c92ad30b1d0cfa507fe9d9a48b14c4046ad  index2/meta.smoosh
febea93b2c7d75eb566bf43de2488821733303f5  index2/version.bin
```

drcrallen commented:
Cool, for binary equivalence and approval from at least two committers, please don't consider me blocking (but since I didn't do a full review I can't be one of the plus ones). I'm excited to see this move forward more.

gianm commented:
thanks even for a partial review @drcrallen, much appreciated.

jon-wei commented:
Ran a binary equivalence check using IndexPersistBenchmark, equivalence is confirmed.

The legacy merger orders columns differently from the V9 merger, which is why the hashes differ.

Earlier I did catch an issue from this check, where this patch caused a fully null column to be erroneously written to the segment, this is fixed in a new commit, with a new test added for that case.

patch
```
legacy merger, with rollup
ed117be06cd41491dbfe50ce7acf2c35f436cbdb  1473468455019-0/00000.smoosh
124c1000535ad7b47c0eb44ea4f5ea941d51e72f  1473468455019-0/meta.smoosh
febea93b2c7d75eb566bf43de2488821733303f5  1473468455019-0/version.bin

legacy merger, no rollup
ab59235f327ea608a8011d6c23b2b54c486581bd  1473468461807-0/00000.smoosh
884728fdad91189da23609ab414a9990bb940979  1473468461807-0/meta.smoosh
febea93b2c7d75eb566bf43de2488821733303f5  1473468461807-0/version.bin

v9 merger, with rollup
f4209f50539e3ed5ca59d5d25b476802b8695275  1473468468730-0/00000.smoosh
bf54f4a77db631d8eb6dc59068462b26e01e3784  1473468468730-0/meta.smoosh
febea93b2c7d75eb566bf43de2488821733303f5  1473468468730-0/version.bin

v9 merger, no rollup
7006cc8e01465603e1186fb99ad5023dbfdadaad  1473468475225-0/00000.smoosh
418b21c8ce2bb5b6dbaf3406894bab398c0d9c17  1473468475225-0/meta.smoosh
febea93b2c7d75eb566bf43de2488821733303f5  1473468475225-0/version.bin
```

master
```
legacy merger, with rollup
ed117be06cd41491dbfe50ce7acf2c35f436cbdb  1473468651547-0/00000.smoosh
124c1000535ad7b47c0eb44ea4f5ea941d51e72f  1473468651547-0/meta.smoosh
febea93b2c7d75eb566bf43de2488821733303f5  1473468651547-0/version.bin

legacy merger, no rollup
ab59235f327ea608a8011d6c23b2b54c486581bd  1473468658437-0/00000.smoosh
884728fdad91189da23609ab414a9990bb940979  1473468658437-0/meta.smoosh
febea93b2c7d75eb566bf43de2488821733303f5  1473468658437-0/version.bin

v9 merger, with rollup
f4209f50539e3ed5ca59d5d25b476802b8695275  1473468665298-0/00000.smoosh
bf54f4a77db631d8eb6dc59068462b26e01e3784  1473468665298-0/meta.smoosh
febea93b2c7d75eb566bf43de2488821733303f5  1473468665298-0/version.bin

v9 merger, no rollup
7006cc8e01465603e1186fb99ad5023dbfdadaad  1473468671867-0/00000.smoosh
418b21c8ce2bb5b6dbaf3406894bab398c0d9c17  1473468671867-0/meta.smoosh
febea93b2c7d75eb566bf43de2488821733303f5  1473468671867-0/version.bin
```

jon-wei commented:
@drcrallen thanks for the review, it was very helpful!

gianm commented:
👍  after travis

gianm commented:
@jon-wei, looks like you used some jdk8-isms in a test file:

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.2:testCompile (default-testCompile) on project druid-processing: Compilation failure: Compilation failure:
[ERROR] /home/travis/build/druid-io/druid/processing/src/test/java/io/druid/segment/IndexMergerTest.java:[2021,24] cannot find symbol
[ERROR] symbol:   method sort(<nulltype>)
[ERROR] location: variable expectedColumnNames of type java.util.List<java.lang.String>
[ERROR] /home/travis/build/druid-io/druid/processing/src/test/java/io/druid/segment/IndexMergerTest.java:[2022,22] cannot find symbol
[ERROR] symbol:   method sort(<nulltype>)
[ERROR] location: variable actualColumnNames of type java.util.List<java.lang.String>
[ERROR] /home/travis/build/druid-io/druid/processing/src/test/java/io/druid/segment/IndexMergerTest.java:[2028,22] cannot find symbol
[ERROR] symbol:   method sort(<nulltype>)
[ERROR] location: variable expectedFilenames of type java.util.List<java.lang.String>
[ERROR] /home/travis/build/druid-io/druid/processing/src/test/java/io/druid/segment/IndexMergerTest.java:[2029,20] cannot find symbol
[ERROR] symbol:   method sort(<nulltype>)
[ERROR] location: variable actualFilenames of type java.util.List<java.lang.String>
[ERROR] -> [Help 1]
```

jon-wei commented:
hm, fixing that now...

gianm commented:
thx @jon-wei
